Sources/System/Sockets/SocketDescriptor+IPv4Address.swift

```diff
@@ -0,0 +1,139 @@
+/*
+ This source file is part of the Swift System open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift System project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+*/
+
+extension SocketDescriptor {
+  @frozen
+  public struct IPv4Address: RawRepresentable {
+    public var rawValue: CInterop.SockAddrIn
+
+    public init(rawValue: CInterop.SockAddrIn) {
+      self.rawValue = rawValue
+    }
+
+    public init?(_ address: SocketDescriptor.Address) {
+      guard address.domain == Domain.ipv4 else { return nil }
+      let value: CInterop.SockAddrIn? = address.withUnsafeBytes { buffer in
+        guard buffer.count >= MemoryLayout<CInterop.SockAddrIn>.size else {
+          return nil
+        }
+        return buffer.baseAddress!.load(as: CInterop.SockAddrIn.self)
+      }
+      guard let value = value else { return nil }
+      self.rawValue = value
+    }
+  }
+}
+
+extension SocketDescriptor.Address {
+  public init(_ address: SocketDescriptor.IPv4Address) {
+    self = Swift.withUnsafeBytes(of: address) { buffer in
+      SocketDescriptor.Address(buffer)
+    }
+  }
+}
+
+extension SocketDescriptor.IPv4Address {
+  public var port: Port {
+    get { Port(CInterop.InPort(_networkOrder: rawValue.sin_port)) }
+    set { rawValue.sin_port = newValue.rawValue._networkOrder }
+  }
+
+  @frozen
+  public struct Port: RawRepresentable, ExpressibleByIntegerLiteral {
+    /// The port number, in host byte order.
+    public var rawValue: CInterop.InPort
+
+    public init(_ value: CInterop.InPort) {
+      self.rawValue = value
+    }
+
+    public init(rawValue: CInterop.InPort) {
+      self.init(rawValue)
+    }
+
+    public init(integerLiteral value: CInterop.InPort) {
+      self.init(value)
+    }
+  }
+}
+
+extension SocketDescriptor.IPv4Address.Port: CustomStringConvertible {
+  public var description: String {
+    "\(rawValue)"
+  }
+}
+
+extension SocketDescriptor.IPv4Address {
+  public var address: Address {
+    get {
+      let value = CInterop.InAddrT(_networkOrder: rawValue.sin_addr.s_addr)
+      return Address(rawValue: value)
+    }
+    set {
+      rawValue.sin_addr.s_addr = newValue.rawValue._networkOrder
+    }
+  }
+
+  @frozen
+  public struct Address: RawRepresentable {
+    /// The raw internet address value, in host byte order.
+    public var rawValue: CInterop.InAddrT
+
+    public init(rawValue: CInterop.InAddrT) {
+      self.rawValue = rawValue
+    }
+  }
+}
+
+extension SocketDescriptor.IPv4Address.Address: CustomStringConvertible {
+  public var description: String {
+    _inet_ntop()
+  }
+
+  internal func _inet_ntop() -> String {
```

Comment:
Should this be `throws`, and then `description` would fatal error?

Reply:
I'm pretty sure this will only report a failure if we messed up the invocation. 

```
     The inet_ntop() call fails if:

     [EAFNOSUPPORT]     *src was not an AF_INET or AF_INET6 family address.

     [ENOSPC]           size was not large enough to store the presentation form of the address.
```

We always pass AF_INET (or AF_INET6), and we always allocate enough memory.